wolf/media/w_openal.hpp

```diff
@@ -14,8 +14,156 @@
 #include <AL/alext.h>
 #include <DISABLE_ANALYSIS_END>
 
+#include <iostream>
+#include <concepts>
+
 namespace wolf::media {
 
+namespace internal {
+
+/**
+ * @brief spsc fixed-size continuous ring buffer. read/write in chunks.
+ *
+ * @tparam T an integral type. buffer read/write happens with memcopy.
+ */
+template <typename T>
+    requires std::is_integral_v<T>
+class w_ring_buffer {
```

Comment:
Please move w_ring_buffer.hpp as a stand-alone header within "wolf::system" namespace and then include it here.

Reply:
done. with a slight change of name from `w_ring_buffer` to `w_spsc_ring_buffer` to differentiate from normal (usually mpmc) ring buffer as it's now a general utility (and not internal) in the library.